erpnext/healthcare/doctype/inpatient_medication_entry/inpatient_medication_entry.py

```diff
@@ -0,0 +1,272 @@
+					entry.idx, get_link_to_form(entry.against_imo)))
+
+			if is_completed:
+				frappe.throw(_('Row {0}: This Medication Order is already marked as completed').format(
+					entry.idx))
+
+	def on_cancel(self):
+		self.cancel_stock_entries()
+		self.update_medication_orders(on_cancel=True)
+
+	def process_stock(self):
+		allow_negative_stock = frappe.db.get_single_value('Stock Settings', 'allow_negative_stock')
+		if not allow_negative_stock:
+			self.check_stock_qty()
+
+		return self.make_stock_entry()
+
+	def update_medication_orders(self, on_cancel=False):
+		orders, order_entry_map = self.get_order_entry_map()
+		# mark completion status
+		is_completed = 1
+		if on_cancel:
+			is_completed = 0
+
+		frappe.db.sql("""
+			UPDATE `tabInpatient Medication Order Entry`
+			SET is_completed = %(is_completed)s
+			WHERE name IN %(orders)s
+		""", {'orders': orders, 'is_completed': is_completed})
+
+		# update status and completed orders count
+		for order, count in order_entry_map.items():
+			medication_order = frappe.get_doc('Inpatient Medication Order', order)
+			completed_orders = flt(count)
+			current_value = frappe.db.get_value('Inpatient Medication Order', order, 'completed_orders')
+
+			if on_cancel:
+				completed_orders = flt(current_value) - flt(count)
+			else:
+				completed_orders = flt(current_value) + flt(count)
+
+			medication_order.db_set('completed_orders', completed_orders)
+			medication_order.set_status()
+
+	def get_order_entry_map(self):
+		# for marking order completion status
+		orders = []
+		# orders mapped
+		order_entry_map = dict()
+
+		for entry in self.medication_orders:
+			orders.append(entry.against_imoe)
+			parent = entry.against_imo
+			if not order_entry_map.get(parent):
+				order_entry_map[parent] = 0
+
+			order_entry_map[parent] += 1
+
+		return orders, order_entry_map
+
+	def check_stock_qty(self):
+		from erpnext.stock.stock_ledger import NegativeStockError
+
+		drug_availability = dict()
+		for d in self.medication_orders:
+			if not drug_availability.get(d.drug_code):
+				drug_availability[d.drug_code] = 0
+			drug_availability[d.drug_code] += flt(d.dosage)
+
+		for drug, dosage in drug_availability.items():
+			available_qty = get_latest_stock_qty(drug, self.warehouse)
+
+			# validate qty
+			if flt(available_qty) < flt(dosage):
+				frappe.throw(_('Quantity not available for {0} in warehouse {1}').format(
+					frappe.bold(drug), frappe.bold(self.warehouse))
+					+ '<br><br>' + _('Available quantity is {0}, you need {1}').format(
+					frappe.bold(available_qty), frappe.bold(dosage))
+					+ '<br><br>' + _('Please enable Allow Negative Stock in Stock Settings or create Stock Entry to proceed.'),
+					NegativeStockError, title=_('Insufficient Stock'))
+
+	def make_stock_entry(self):
+		stock_entry = frappe.new_doc('Stock Entry')
+		stock_entry.stock_entry_type = 'Material Issue'
```

Comment:
stock_entry_type should not be hard-coded, as it is a master. You can set purpose as 'Material Issue'. ON save of stock entry, it will automatically set stock_entry_type based on purpose.